addons/payment/tests/test_payment_transaction.py

```diff
@@ -80,3 +81,29 @@ def test_refund_transaction_values(self):
             msg="The amount of the refund transaction should be the negative value of the amount "
                 "to refund."
         )
+
+    @mute_logger('odoo.addons.payment.models.payment_transaction')
+    def test_update_state_from_wrong_state(self):
+        tx = self._create_transaction('redirect', state='done')
+        self.assertFalse(
+            tx._update_state(
+                allowed_states=['draft', 'pending', 'authorized'],
+                target_state='cancel',
+                state_message=None
+            ),
+            msg="Transaction without extra allowed states should not be able to change state"
+                "if in wrong state."
+            )
```

Comment:
This test is useful and welcome, but I find that it verifies an assertion that is only very indirectly related to the behavior we want to test.

Here, you ensure that: 1. calling `_updated_state` returns a falsy value; 2. that value is falsy because it's an empty recordset; 3. the recordset is empty because `classify_by_state` returned an empty recordset too; 4. `classify_by_state` returned an empty recordset because of the `filtered` on `allowed_states`.

Although it is correct, I believe it would be clearer to directly test that the transaction whose state you try to update stays in the same state. So, something like
```py
tx._update_state(...)
self.assertEqual(tx.state, 'done')
```